### PR TITLE
Cache packages folder builds to cut down the build time

### PR DESCRIPTION
**Pull request by alshakero**
#### Changes proposed in this Pull Request

In an attempt to cut down our CI times, I did some investigation looking for low hanging fruits. I found out that `build-packages` job takes 3-5 minutes. I then calculated how common is it to modify the packages directory. And turns out we only modify it 84 times in every 1000 commits. ~8.4% of the time.

Here is how I counted this (I hope it's right):
```sh
positive_commits=0

for i in `seq 1000 2`
do
   git --no-pager diff --name-only HEAD~$i HEAD~$(expr $i - 1) | grep "packages/" -q && (( positive_commits=positive_commits+1 ))
done

echo $positive_commits # it's 84
```

This doesn't mean that we don't need the packages ready and built for every build. We need them, but we don't them *fresh*™. Let's cache them. 

#### It's a two part story, Locally and in CI, and they interplay

**Locally:** 

1. Assume all files in packages folder (minus `node_modules`, `dist`) play a role in the build result. 
2. Calculate the checksum for these files and store it in a git-ignored file (packages/packages_checksum.txt).
3. Create a `build-packages:if-needed` command that calculates the fresh checksum of the packages directory and compare it against the stored one. If they match, it means no build is needed. And since the checksu, file is gitignored, it's guaranteed that you built the packages locally. It's worth nothing that this checksum generator should incorporate `.nvmrc` file into the hash. To make sure  packages are built under the same Node version. 
5. This saves 20-40s. Profit. 

##### In the CI
Using the same checksum + .nvmrc as cache keys, cache the `packages` folder as is. When it exists, it is guaranteed to be fresh, since the key is the checksum of the source files. Nothing is returned from CircleCI cache if the cache key doesn't exist. Nevertheless, `yarn build` will double check packages folder freshness, because the rules from "Locally" above apply in the CI. 

##### Invalidation
Calling `yarn clean`, `yarn clean:packages` will delete the checksum file and this the packages will be rebuilt. Also, calling `yarn build-packages` will always build the packages, all these checks only apply to `yarn build-packages:if-needed". Frankly, I don't see a use case for `yarn build-packages`, but caching can always get tricky. 

#### Testing instructions

1. Run `yarn build`
2. It should build packages.
3. Run `yarn build`
4. It should say "Packages are already freshly built". 
5. Modify any file in the packages folder that is not in `dist` or `node_modules`.
6. Run `yarn build`. It should rebuild the packages.

### CODE REVIEW

**Review by scinos — CHANGES_REQUESTED** [review]
A few notes:

* `yarn build-packages` also builds `apps/`, not only `packages/`. We should include them in the checksum and any other computation
* The checksum also has to include dependencies. Some of them will live in `./packages/**/node_modules` or `./apps/**/node_modules` but most of them will be in `./node_modules`. A proxy may be to include `package.json` and `./yarn.lock` in the hash, but it is not 100% guaranteed (a package could be importing a module that is not declared in its `package.json`).

File: bin/do-packages-need-a-rebuild.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+const checksumFilePath = path.join( __dirname, '..', 'packages', 'packages-checksum.txt' );
+
+// map exit code to answer for less confusion
+const ANSWERS = {
+	YES: 0,
+	NO: 1,
+};
+
+async function shouldRebuild() {
```
Comment (blowery) [inline bin/do-packages-need-a-rebuild.js:14]: we should also rebuild if anything in the compilation chain has changed. Different babel version, babel plugin versions, typescript versions, etc etc.

Comment (blowery) [reply bin/do-packages-need-a-rebuild.js:14]: a decent proxy for that might be a combo hash of the root yarn.lock and root babel config

**Comment by simison** [comment]
> yarn build-packages also builds apps/, not only packages/. We should include them in the checksum and any other computation

`apps` and `packages` could have their separate cache-busting though.

**Comment by alshakero** [comment]
Thank you @scinos and @blowery for the reviews! I addressed all issues.

**Comment by alshakero** [comment]
I feel that this change, albeit makes a considerable performance difference, it overcomplicates a bit and might make maintenance harder. I'll close it and we can look at for inspo later.

**Pull request closed without merging**